bug fix: warning for old patches

rd4398 opened the pull request:
Fixes: https://github.com/python-wheel-build/fromager/issues/254

Also, I have used the test case suggested by @shubhbapna which was failing

shubhbapna commented on src/fromager/overrides.py:
```diff
@@ -83,6 +83,18 @@ def patches_for_source_dir(
+    It takes in patches directory and requirement as input.
+    The output will be a list of all directories containing patches for that requirement
+    """
+    glob_string = req + "*"
+    return sorted((patches_dir).glob(glob_string))
```
```suggestion
    return sorted((patches_dir).glob(f"{req}*"))
```

rd4398 replied:
Done

shubhbapna commented on src/fromager/overrides.py:
```diff
@@ -83,6 +83,18 @@ def patches_for_source_dir(
 
 
+def patches_for_req(
+    patches_dir: pathlib.Path, req: str
```
```suggestion
    patches_dir: pathlib.Path, req_name: str
```
since you just pass only the req name so maybe this var name might be better

rd4398 replied:
Done

shubhbapna commented on src/fromager/overrides.py:
```diff
@@ -83,6 +83,18 @@ def patches_for_source_dir(
 
+def patches_for_req(
+    patches_dir: pathlib.Path, req: str
+) -> typing.Iterable[pathlib.Path]:
```
I think a better return type would be a `list` since we are using `len` in the warn function and `len` is not necessarily part of an iterable. @tiran might be able to confirm?

rd4398 replied:
Okay, I have set the return type to list now. I will change it again if required.

shubhbapna commented on src/fromager/sources.py:
```diff
@@ -311,17 +309,19 @@ def patch_source(
+    req_patches = overrides.patches_for_req(patches_dir, req_name)
+    # Check if old patches exist and current patch not in the list
+    if len(req_patches) > 0 and patch_path not in req_patches:
+        logger.warning(f"{req_name}: old patches exist but will not be applied")
```
Probably want to log the version of the patches not being applied (logging the patches directory might be enough)

rd4398 replied:
Done

shubhbapna reviewed (changes requested):
Could you also add a test for a case where no warning should be applied

rd4398 commented:
Okay, making changes!

dhellmann commented on src/fromager/overrides.py:
```diff
@@ -83,6 +83,15 @@ def patches_for_source_dir(
+    It takes in patches directory and requirement as input.
+    The output will be a list of all directories containing patches for that requirement
+    """
+    return sorted((patches_dir).glob(f"{req_name}*"))
```
We have some packages that start with a common prefix, so this is going to return more than it should. For example, we have `instructlab` and `instructlab-sdg`. Any patches for `instructlab-sdg` will appear to be related to `instructlab`. The directory naming convention includes a `-` between the package name and version, so you could add that to the pattern to fix this.

There will also be patches in a directory just named for the requirement, without any version. For that case, you could just check for that directory to exist at all and include it if it is there without using a glob pattern.

The `patches_for_source_dir()` function takes the source directory, instead of a requirement, because the source directory doesn't have to match the requirement name. Requirements like `instructlab.sdg` and `instructlab-sdg` are equivalent. It looks like you're expecting the prefix of the source directory name, without any version information. This function might be clearer if it took the full source directory name and did its own parsing of that name to strip the version string, instead of having the caller do that step.

dhellmann commented on src/fromager/sources.py:
```diff
@@ -311,17 +309,21 @@ def patch_source(
+    # Check if old patches exist and current patch not in the list
+    if len(req_patches) > 0:
+        string_to_search = "\t".join(str(p) for p in req_patches)
+        if source_root_dir.name not in string_to_search:
```
I'm worried that relying on substrings is going to be brittle because of the name collisions I mentioned in the earlier comment.

If you build 2 sets, one containing all patches for all versions of a package and the other containing only patches for the current version of the package, you could use set.difference() to find the items that are available to be applied that won't be applied and then emit the warnings.

rd4398 replied:
How can I build a set for patches for current version? Should I add a method in overrides doing something like:

```
def patches_directory_for_current_version(patches_dir: pathlib.Path, source_dir_name: str) -> pathlib.Path | None:
    """
    This method will return patches directory for current version
    """
    return patches_dir / source_dir_name if (patches_dir / source_dir_name).exists() else None
```
Or is there  a better way? I am not sure

rd4398 commented:
After, doing a rebase and understanding Shubh's changes for generic patch, I have changed the logic for warning of old patches which now includes a regex to filter packages and avoid conflicts for packages like `instructlab `and `instructlab-sdg`
As per Doug's comment in https://github.com/python-wheel-build/fromager/issues/254:
`Remember, the goal is to report patches related to a package that will not be applied, but only when no patches will be applied.`
I think set difference is not needed and a simple flag in `patch_source` is enough. I have added test cases for different scenarios as well.
Does it make sense?
Let me know if any changes are required

dhellmann commented on src/fromager/sources.py:
```diff
@@ -326,17 +336,30 @@ def _apply_patch(patch: pathlib.Path, source_root_dir: pathlib.Path):
+    if len(req_patches) > 0:
+        # Set up regex to filter out unwanted patches.
+        # Example: instructlab and instructlab-sdg
+        pattern = re.compile(rf"^{req_name}(-v?\d+\.\d+\.\d+)?$")
```
vllm will have 4 numerical parts to its version, and we have some packages that end with version `.post1`. If there are unversioned patches, in a directory that's the same as req_name, then those would have been applied and this function shouldn't be called, so I think you can ignore that case.

I think that's something like this:

```suggestion
        pattern = re.compile(rf"^{req_name}-v?(\d+\.)+\d+)")
```

It would be good to pull this regex out of the function and write some unit tests using known package names and versions that we have.

rd4398 replied:
Okay, I have added tests for regex and created a helper for it

dhellmann commented on src/fromager/sources.py:
```diff
@@ -326,17 +336,30 @@ def _apply_patch(patch: pathlib.Path, source_root_dir: pathlib.Path):
+        if current_path in filtered_patches:
+            filtered_patches.remove(current_path)
+        if len(filtered_patches) > 0:
+            logger.warning(f"{req_name}: old patches exist but will not be applied")
```
They might be patches for a newer version, so the message shouldn't assume which one is the newer/older. It would be good to log each patch name that is being ignored.

rd4398 replied:
Okay, I am now logging` filtered_patches`. 
Should I use `iterdir()` to log individual patch files?

rd4398 commented on src/fromager/overrides.py:
```diff
@@ -83,6 +83,17 @@ def patches_for_source_dir(
+def patches_for_req(patches_dir: pathlib.Path, dir_name: str) -> list[pathlib.Path]:
+    """
+    This function will return patches for a specific requirement.
+    It takes in patches directory and requirement as input.
```
Okay, changed it to `get_patches()`
Is it more generic? Should I change it?

dhellmann replied:
The docstring is out of date. What is `dir_name` expected to be?

This function is returning all patches for all versions of a dependency, starting from a source directory, right? If that's the case, please change `dir_name` to `source_root_dir` to match the value used elsewhere. Naming it consistently helps the caller understand what it is.

Since the function doesn't return patch filenames, it shouldn't be called `get_patches`. Maybe `get_patch_directories`?

rd4398 commented:
I have made changes as per the suggestions. Let me know if any other changes are required

dhellmann commented on src/fromager/sources.py:
```diff
@@ -326,17 +336,36 @@ def _apply_patch(patch: pathlib.Path, source_root_dir: pathlib.Path):
+
+
+# Helper method to filter the unwanted patches using a regex
+def _filter_patches(patches: list[pathlib.Path], req_name: str) -> list[pathlib.Path]:
```
This function needs a more descriptive name. 

Maybe this filtering should be done in the function in overrides?

rd4398 replied:
Done. Also shifted it to overrides.py

dhellmann commented on src/fromager/sources.py:
```diff
@@ -326,17 +336,36 @@ def _apply_patch(patch: pathlib.Path, source_root_dir: pathlib.Path):
+        # If filtered patches contain patches for current version, remove the path because we dont want to warn
+        if current_path in filtered_patches:
+            filtered_patches.remove(current_path)
+        if len(filtered_patches) > 0:
```
Iterating over an empty list is OK, you don't need this test.

rd4398 replied:
Okay

dhellmann commented on src/fromager/sources.py:
```diff
@@ -326,17 +336,36 @@ def _apply_patch(patch: pathlib.Path, source_root_dir: pathlib.Path):
+    # Get all the patches for that requirement
+    patches = overrides.get_patches(patches_dir, source_root_dir.name)
+
+    if len(patches) > 0:
```
This test isn't needed. If the list is empty, no messages will be emitted.

dhellmann commented on src/fromager/overrides.py:
```diff
@@ -83,6 +84,32 @@ def patches_for_source_dir(
+    patches_dir: pathlib.Path, source_root_dir: pathlib.Path
+) -> list[pathlib.Path]:
+    """
+    This function will return patches for a specific requirement.
```
```suggestion
    This function will return directories that may contain patches for a specific requirement.
```

rd4398 replied:
Okay, done!

dhellmann commented on src/fromager/overrides.py:
```diff
@@ -83,6 +84,32 @@ def patches_for_source_dir(
+) -> list[pathlib.Path]:
+    """
+    This function will return patches for a specific requirement.
+    It takes in patches directory and requirement as input.
```
```suggestion
    It takes in patches directory and a source root directory as input.
```

rd4398 replied:
done!

dhellmann reviewed (commented):
2 nits on that docstring, and then I think this is ready

dhellmann reviewed (commented):
This version looks good to me. @shubhbapna had comments, so let's get him to look again before we merge it.

shubhbapna commented on tests/test_sources.py:
```diff
@@ -179,3 +165,139 @@ def test_patch_sources_apply_only_unversioned(
+        patches_dir / "deepspeed-deep-2.0.0",
+        patches_dir / "deepspeed-v2.3.0.post1",
+        patches_dir / "deepspeed-v5.5.5",
+        patches_dir / "deepspeed-3.4.4",
+        patches_dir / "deepspeed-sdg-3.4.4",
+    ]
+
+    expected = [
+        patches_dir / "deepspeed-1.1.0",
+        patches_dir / "deepspeed-v2.3.0.post1",
+        patches_dir / "deepspeed-v5.5.5",
+        patches_dir / "deepspeed-3.4.4",
+    ]
+
+    actual = overrides._filter_patches_based_on_req(lst, req_name)
+    assert len(expected) == len(actual)
+    assert expected == actual
+
+
+def test_regex_for_vllm(tmp_path: pathlib.Path):
+    req_name = "vllm"
+    patches_dir = tmp_path / "patches_dir"
+    patches_dir.mkdir()
+
+    lst = [
+        patches_dir / "vllm-1.1.0.9",
+        patches_dir / "vllm-llm-2.1.0.0",
+        patches_dir / "vllm-v2.3.5.0.post1",
+        patches_dir / "vllm-v5.5.5.1",
+    ]
+
+    expected = [
+        patches_dir / "vllm-1.1.0.9",
+        patches_dir / "vllm-v2.3.5.0.post1",
+        patches_dir / "vllm-v5.5.5.1",
+    ]
+
+    actual = overrides._filter_patches_based_on_req(lst, req_name)
+    assert len(expected) == len(actual)
+    assert expected == actual
```
These tests should be in `test_overrides.py`

rd4398 replied:
okay, done!

shubhbapna commented on src/fromager/sources.py:
```diff
@@ -326,17 +335,21 @@ def _apply_patch(patch: pathlib.Path, source_root_dir: pathlib.Path):
-            f"{req_name}: patches for version {existing_version!s} exist but will not be applied"
-        )
+    # Filter the patches using regex
+    patch_directories = overrides.get_patch_directories(patches_dir, source_root_dir)
+    current_path = source_root_dir.parent.joinpath("patches_dir", source_root_dir.name)
+    # If filtered patches contain patches for current version, remove the path because we dont want to warn
+    if current_path in patch_directories:
+        patch_directories.remove(current_path)
```
I don't think filtered patches will contain the current version since you are only calling this function if `has_applied` is false which implies that were no patches directory for the current version.

You can add a docstring for this helper to clarify this point so that any other caller can use it correctly

rd4398 replied:
Okay, I added it as docstring

rd4398 commented:
The latest version is ready for review

dhellmann reviewed (approved):
This looks good, thanks for those updates!

shubhbapna reviewed (approved):
LGTM
